File: app/src/main/java/com/alexkn/syntact/domain/usecase/play/SolvableItemRepository.java
package com.alexkn.syntact.domain.usecase.play;

import android.content.Context;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.work.Data;
import androidx.work.ExistingWorkPolicy;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

import com.alexkn.syntact.dataaccess.common.AppDatabase;
import com.alexkn.syntact.dataaccess.dao.BucketDao;
import com.alexkn.syntact.dataaccess.dao.SolvableItemDao;
import com.alexkn.syntact.domain.model.Bucket;
import com.alexkn.syntact.domain.model.SolvableItem;
import com.alexkn.syntact.domain.model.cto.SolvableTranslationCto;
import com.alexkn.syntact.restservice.SolvableItemRemoteRepository;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Maybe;

@Singleton
public class SolvableItemRepository {

    private static final String TAG = SolvableItemRepository.class.getSimpleName();

    private SolvableItemRemoteRepository solvableItemRemoteRepository;

    private SolvableItemDao solvableItemDao;

    private BucketDao bucketDao;

    @Inject
    SolvableItemRepository(SolvableItemRemoteRepository solvableItemRemoteRepository, Context context) {

        this.solvableItemRemoteRepository = solvableItemRemoteRepository;
        solvableItemDao = AppDatabase.getDatabase(context).solvableItemDao();
        bucketDao = AppDatabase.getDatabase(context).bucketDao();
    }

    public LiveData<List<SolvableTranslationCto>> getSolvableTranslations(Long bucketId) {

        return solvableItemDao.getSolvableTranslationsDueBefore(bucketId, Instant.now());
    }

    public Maybe<SolvableTranslationCto[]> getNextSolvableTranslations(Long bucketId, Instant time, int count) {

        Maybe<SolvableTranslationCto[]> translations = solvableItemDao.getNextTranslationDueBefore(bucketId, time, count);

        return translations.flatMap(items -> {
            if (items.length == count) {
                Log.i(TAG, "All translations present");
                return Maybe.just(items);
            } else {
                Log.i(TAG, "Translations missing, fetching remote");
                Bucket bucket = bucketDao.find(bucketId);
                long maxId = solvableItemDao.getMaxId(bucketId);
                List<SolvableTranslationCto> solvableTranslationCtos = solvableItemRemoteRepository.fetchNewTranslations(bucket, time, maxId, count);
                //TODO Service to return Maybe
                return Maybe.just(solvableTranslationCtos.toArray(new SolvableTranslationCto[0]));
            }
        });
    }

    public void fetchSolvableItems(long bucketId, Instant startTime) {

        Data data = new Data.Builder().putLong("bucketId", bucketId).putLong("timestamp", startTime.toEpochMilli()).build();

        OneTimeWorkRequest workRequest = new OneTimeWorkRequest.Builder(FetchPhrasesWorker.class).setInputData(data).build();
        WorkManager.getInstance().enqueueUniqueWork(FetchPhrasesWorker.class.getName(), ExistingWorkPolicy.KEEP, workRequest);
    }

    public void solvePhrase(SolvableTranslationCto solvableTranslation) {

        SolvableItem solvableItem = solvableTranslation.getSolvableItem();
        Log.i(TAG, "Solving Phrase " + solvableItem.getText());

        int performanceRating = 3;
        float easiness = solvableItem.getEasiness();
        easiness += -0.80 + 0.28 * performanceRating + 0.02 * performanceRating * performanceRating;

        int consecutiveCorrectAnswers = solvableItem.getConsecutiveCorrectAnswers() + 1;

        double daysToAdd = 6 * Math.pow(easiness, consecutiveCorrectAnswers - 1d);
        Instant nextDueDate = Instant.now().plus(Math.round(daysToAdd), ChronoUnit.DAYS);

        solvableItem.setEasiness(easiness);
        solvableItem.setConsecutiveCorrectAnswers(consecutiveCorrectAnswers);
        solvableItem.setNextDueDate(nextDueDate);
        solvableItem.setLastSolved(Instant.now());
        solvableItem.setTimesSolved(solvableItem.getTimesSolved() + 1);

        solvableItemDao.update(solvableItem);
    }
}

File: app/src/main/java/com/alexkn/syntact/core/repository/SolvableItemRepository.kt
package com.alexkn.syntact.core.repository

import android.util.Log
import androidx.lifecycle.LiveData
import com.alexkn.syntact.app.TAG
import com.alexkn.syntact.data.dao.SolvableItemDao
import com.alexkn.syntact.data.model.SolvableTranslationCto
import java.time.Instant
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.roundToLong


@Singleton
class SolvableItemRepository @Inject constructor(
        private val solvableItemDao: SolvableItemDao
) {

    private val maxPerformanceRating = 5.0

    fun getSolvableTranslations(bucketId: Long): LiveData<List<SolvableTranslationCto>> = solvableItemDao.getSolvableTranslations(bucketId)

    suspend fun findNextSolvableItem(deckId: Long, time: Instant): SolvableTranslationCto? {

        val newItems = findNewItems(deckId, 20 - findItemsSolvedOnDay(deckId, time).size)
        if (newItems.isNotEmpty()) return newItems[0]
        val reviewItems = findItemsDueForReview(deckId, time)
        return if (reviewItems.isNotEmpty()) reviewItems[0] else null
        //TODO shuffle, new first, review first
    }

    suspend fun findItemsDueForReview(deckId: Long, time: Instant): List<SolvableTranslationCto> {

        val endOfDay = time.atZone(ZoneId.systemDefault()).withHour(23).withMinute(59).withSecond(59).toInstant()
        return solvableItemDao.findSolvedItemsDueBefore(deckId, endOfDay)
    }

    suspend fun findNewItems(deckId: Long, limit: Int): List<SolvableTranslationCto> {

        return solvableItemDao.findUnsolvedItems(deckId, limit)
    }

    suspend fun findItemsSolvedOnDay(deckId: Long, time: Instant): List<SolvableTranslationCto> {

        val from = time.atZone(ZoneId.systemDefault()).withHour(23).withMinute(59).withSecond(59).toInstant()
        val to = time.atZone(ZoneId.systemDefault()).withHour(0).withMinute(0).withSecond(0).toInstant()
        return solvableItemDao.findItemsSolvedBetween(deckId, from, to)
    }

    suspend fun markPhraseCorrect(solvableTranslation: SolvableTranslationCto, scoreRatio: Double) {

        val solvableItem = solvableTranslation.solvableItem

        Log.d(TAG, "markPhraseCorrect: $solvableItem (Before)")

        val performanceRating = maxPerformanceRating * scoreRatio
        var easiness = solvableItem.easiness
        easiness += (-0.80 + 0.28 * performanceRating + 0.02 * performanceRating * performanceRating).toFloat()

        val daysToAdd = 6 * easiness.toDouble().pow(solvableItem.consecutiveCorrectAnswers - 1.0)
        val consecutiveCorrectAnswers = solvableItem.consecutiveCorrectAnswers + 1

        val nextDueDate = Instant.now().plus(daysToAdd.roundToLong(), ChronoUnit.DAYS)

        solvableItem.easiness = easiness
        solvableItem.consecutiveCorrectAnswers = consecutiveCorrectAnswers
        solvableItem.nextDueDate = nextDueDate
        solvableItem.lastSolved = Instant.now()
        solvableItem.timesSolved = solvableItem.timesSolved + 1

        solvableItemDao.update(solvableItem)
        Log.d(TAG, "markPhraseCorrect: $solvableItem (Updated)")
    }

    suspend fun markPhraseIncorrect(solvableTranslation: SolvableTranslationCto, scoreRatio: Double) {

        val solvableItem = solvableTranslation.solvableItem

        Log.d(TAG, "markPhraseIncorrect: $solvableItem (Before)")

        val performanceRating = maxPerformanceRating * scoreRatio
        var easiness = solvableItem.easiness
        easiness += (-0.80 + 0.28 * performanceRating + 0.02 * performanceRating * performanceRating).toFloat()
        easiness = max(easiness, 1.3f)

        val consecutiveCorrectAnswers = 0

        val nextDueDate = Instant.now().plus(1, ChronoUnit.DAYS)

        solvableItem.easiness = easiness
        solvableItem.consecutiveCorrectAnswers = consecutiveCorrectAnswers
        solvableItem.nextDueDate = nextDueDate

        solvableItemDao.update(solvableItem)
        Log.d(TAG, "markPhraseIncorrect: $solvableItem (Updated)")
    }

}
